Add support for custom label components.

przywartya opened the pull request.

czosel commented on addon/templates/components/validated-label.hbs:
```diff
@@ -0,0 +1,7 @@
+{{#if isLabel}}
+	<label class={{config.css.label}} for={{inputId}}>
+  	{{label}}{{#if required}} {{requiredLabel}}{{/if}}
+  </label>
```
Nitpick: Indentation

czosel commented on addon/templates/components/validated-input.hbs:
```diff
@@ -82,7 +83,9 @@
-        <span>{{label}}</span>
+        <span style="display: inline-block;">
+          {{component (if labelComponent labelComponent "validated-label") label=label config=config inputId=inputId}}
+        </span>
```
This would render a `<label>` inside another `<label>` - which is not exactly what we want here, right?
Maybe it would make sense to use this style of checkbox labels instead:
```html
<input type="checkbox" id="coding" name="interest" value="coding">
<label for="coding">Coding</label>
```
(taken from [MDN](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/checkbox))

With this style, it seems that we could use `labelComponent` / `validated-label` exactly in the same way as with the other inputs. And it looks like all that needs to be changed is that the enclosing `<label>` tag has to be removed.

What do you think? :smile:

czosel replied:
Edit: Just realized that this is what `isLabel` is supposed to do. What do you think about using the `label for="..."` syntax instead, so that we don't need `isLabel`?

anehx commented on addon/templates/components/validated-input.hbs:
```diff
@@ -1,8 +1,9 @@
+  {{#if (or label labelComponent)}}
+    <span style="display: inline-block;">
+      {{component (if labelComponent labelComponent "validated-label")
+        isLabel=true label=label config=config inputId=inputId required=required}}
```
I'd prefer always using `{{component labelComponent}}` and just setting the default value of `labelComponent` to `'validated-label'`

przywartya commented:
- there was no `required star` for type checkbox input
- there was no `inputId` for type checkbox input
I have added both those things to checkbox in this PR. @anehx @czosel I don't know why it was not there :face_with_head_bandage: 
@anehx I have added a default :+1: 
@czosel `{{inputId}}` sometimes gives `for="ember366-input-undefined"` (when we don't satisfy `name` argument in `{{validated-input}}`). Tell me if that's okay for you? :wink: :

przywartya commented on addon/templates/components/validated-input.hbs:
```diff
@@ -1,8 +1,8 @@
-        {{label}}{{#if required}} {{requiredLabel}}{{/if}}
-      </label>
+  {{#if (or label (not-eq labelComponent 'validated-label'))}}
+    <span style="display: inline-block;">
```
Ember is wrapping component render in  `<div id="ember..." class="ember-view"> </div>`, but we want our label components to be inline with our input. So I have to add an `inline-block` style :wink:

anehx replied:
Why not just use `tagName: 'label'` on the `validated-label` component and remove it from the template? The `for` and `class` can be easily mapped too. This results in a cleaner DOM since we have way less elements..

przywartya replied:
Cool, I will do that!

czosel commented:
Hi @przywartya,
sorry about the late reply! Code-wise, this looks great now. The only thing this is still missing is some tests - could you look into that? Let me know if I can help :wink:

przywartya commented:
Updated :)

czosel commented:
Looks great! Thanks for your awesome contribution :100:
